### ISSUE TITLE
How to Resize Dcls2d Convolutional Kernel

### ISSUE THREAD

**Issue by CV-ZhangXin** [K-H-Ismail/Dilated-Convolution-with-Learnable-Spacings-PyTorch#6]
Dear Mr. @K-H-Ismail 
Thank you for your outstanding work！
Recently, I want to replace the 3 × 3 standard convolution in YOLOv5 with Dcls2d, but found that the shape of the output is not equal to the shape of the input, when setting kernel_count= 9 and dilated_kernel_size = 1. How should I set the parameters of Dcls2d to fairly compare the 3 × 3 standard convolution operation? Looking forward to your replies.
Best regards.

**Comment by K-H-Ismail** [comment]
Dear @CV-ZhangXin ,

Thank you for your interest!

Perhapse what you want to do is  replacing 
```python 
torch.nn.Conv2d(in_ch, out_ch, kernel_size=3)
``` 
by

```python 
 Dcls2d(in_ch, out_ch, dilated_kernel_size=9, padding=4, kernel_count=3, version='v1') 
```

kernel_count=3 ensures you to have the same number of parameter, dilated_kernel_size could be set independantly from kernel count. However, we noticed that as the dilated_kernel_size the accuracy was better up to a certain point where it saturate.
we used dilated_kernel_size = 17 for our study with ConvNeXt.

The padding is a padding same, so for dilated_kernel_size = d, you should use padding = (d - 1) // 2

Finally, DCLS is faster with depthwise separable convolutions, as we make use of the Implicit gemm method. Maybe you can opt for an architecture using depthwise separable convolutions like YOLOv9, or maybe change the 3x3 convolution to separable ones.

Sincerely,
Ismail.

**Comment by CV-ZhangXin** [comment]
Thank you for your reply!
I would also like to know how to implement a convolution operation with a convolution parameter of 5, i.e. a convolution kernel that is non-square, using Dcls2d.

Best regards.
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "K-H-Ismail/Dilated-Convolution-with-Learnable-Spacings-PyTorch"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年5月10日(星期五) 晚上6:29
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [K-H-Ismail/Dilated-Convolution-with-Learnable-Spacings-PyTorch] How to Resize Dcls2d Convolutional Kernel (Issue #6)





 
Dear @CV-ZhangXin ,
 
Thank you for your interest!
 
Perhapse what you want to do is  replacing
 torch.nn.Conv2d(in_ch, out_ch, kernel_size=3)
 
by
  Dcls2d(in_ch, out_ch, dilated_kernel_size=9, padding=4, kernel_count=3, version='v1') 
 
kernel_count=3 ensures you to have the same number of parameter, dilated_kernel_size could be set independantly from kernel count. However, we noticed that as the dilated_kernel_size the accuracy was better up to a certain point where it saturate.
 we used dilated_kernel_size = 17 for our study with ConvNeXt.
 
The padding is a padding same, so for dilated_kernel_size = d, you should use padding = (d - 1) // 2
 
Finally, DCLS is faster with depthwise separable convolutions, as we make use of the Implicit gemm method. Maybe you can opt for an architecture using depthwise separable convolutions like YOLOv9, or maybe change the 3x3 convolution to separable ones.
 
Sincerely,
 Ismail.

**Comment by CV-ZhangXin** [comment]
Dear Mr.  @K-H-Ismail 
Thank you for your reply!
I would also like to know how to implement a convolution operation with a convolution parameter of 5, i.e. a convolution kernel that is non-square, using Dcls2d.
Best regards.

**Comment by K-H-Ismail** [comment]
Hello @CV-ZhangXin ,
You can do for example for rectangular convs

```python
 Dcls2d(in_ch, out_ch, dilated_kernel_size=(9,5), padding=(4,2), kernel_count=15, version='v1') 
 ```
 The dilated kernel size fixes the maximum rectangle (limit of the kernel) where the kernel weights could move. The kernel count determines how much kernel elements there are inside this kernel.
